### ISSUE TITLE
Container is regenerated for every request

### ISSUE THREAD

**Issue by vhenzl** [wodCZ/nette-dotenv#2]
This library causes that DI container is generated for every request.

The reason for that is `$_SERVER` used to populate container's parameters. It contains client/request specifics values such as `HTTP_USER_AGENT`, `REQUEST_TIME` or `REMOTE_ADDR`.

## Possible solution

`ParameterLoader` should iterate over `$_ENV` instead of `$_SERVER`.

The problem with `$_ENV` is that it works only if [variables_order](http://php.net/manual/en/ini.core.php#ini.variables-order) directive contains `E`. Since it's often set only to `GPCS`,  `$_ENV` could be an empty array and wouldn't be obvious, why it doesn't work.

I mean, `$_ENV` will work for `.env` file without any problem, of course. Problematic are _real environment variables_ which requires having `variables_order` set properly.

`variables_order` can't be set at runtime. That means options are either to modify `php.ini`, define it in CLI command

```
XXX=xxx php -d variables_order=EGPCS index.php
```

or eg. for Apache server enable it in `.htaccess`

```
php_value variables_order EGPCS
```

My point is, it can't be done automatically by this package.

## Possibly better solution

Since `$_SERVER` is always combination of  _S_ and _E_ regardless of `variables_order`, might be more convenient keep using it. In that case, we need to get rid of undesirable values and `ParameterLoader` should provide a way how to whitelisted some parameters.

This could be possibly extended to support [required variables](https://github.com/vlucas/phpdotenv#requiring-variables-to-be-set). Something like

```php
$parametersLoader->getParameters(['OPTIONAL' => false, 'REQUIRED' => true]);
$parametersLoader->getParameters(['OPTIONAL', '!REQUIRED']);
$parametersLoader->getParameters(['?OPTIONAL', 'REQUIRED']);
```

**Comment by wodCZ** [comment]
Hi @vhenzl,
that's a nice catch, thank you for taking time to create this issue!

Depending on system configuration is something I'd like to avoid. In addition to that, I'm afraid that `$_ENV` will still contain these client/request specific variables.

While whitelisting/adding support for required variables sounds good to me, I will please you to give me a day or two to think about more convenient solution.

I think it could be possible to extend configuration parsing or something.

**Comment by vhenzl** [comment]
Nah, `$_ENV` is all right, it contains only _ENVs_.

In the case of whitelisting, maybe would make even more sense to use `getenv` to get allowed values, so you wouldn't need to worry about `$_ENV`  and `$_SERVER` at all. And then it's actually more about specifying of _expected ENVs_ than about whitelisting.

**Comment by vhenzl** [comment]
As I'm thinking about it again, I'm starting to prefer the first solution with `$_ENV`. Just needs to be made clear that _ENVs_ has to be enabled in `variables_order`.

The second option probably only complicates thinks more than is necessary…

**Comment by wodCZ** [comment]
Please have a look at [extension-rewrite](https://github.com/wodCZ/nette-dotenv/tree/extension-rewrite) branch.

I have completely rewritten the package to register a `accessor` service.
It's possible to get any environment variable without defining it at two different places, and it doesn't regenerate container each time.

Only issue with that approach is that the new syntax `@env::get('something')` can't be directly used in strings, like a `DSN` configuration. It's still possible to use `@env` to create new variables in `parameters:` section and use these in `DSN` though.

Please let me know what you think about it.

Thanks!

**Comment by vhenzl** [comment]
Man, that's amazing!

I'll try to integrate it into our application and let you know then.

So far I noticed an issue with variables which have empty value or value equal to `0`. I'll send a PR.

**Comment by wodCZ** [comment]
Merged pr #3, added you to contributors, updated changelog and released v2.0.0.

Thank you!

**Comment by vhenzl** [comment]
Thanks a lot! I upgraded our application to use the new version and everything works like a charm.

A question anyway: Let say I set in `config.neon` a parameter like this:
```yaml
parameters:
    someParam: @env::get('SOME_PARAM')
``` 
Now I can use `%someParam%` to configure services and extensions, but since the parameter is `Nette\DI\Statement` I can't get its value like `$container->getParameters()['someParam']`. And probably there isn't any way, how to turn the _statement_ into a value in code. Am I right?

---

Another thing I had to deal with is fact that ENVs from `.env` are only available after _container_ is created.

That means if my `bootstrap.php` looks like
```php
$configurator = new Nette\Configurator;

$configurator->setDebugMode(getenv('NETTE_DEBUG') === '1');
$configurator->addConfig(__DIR__ . '/config/config.neon');

$container = $configurator->createContainer();
```
`NETTE_DEBUG` can't be defined in `.env` file and must be a _real ENV_. But that's something I can live with...
